### PR TITLE
MGMT-13596: Add PATCH to the Manifests API

### PR DESCRIPTION
**Pull request by paul-maidment**
This is to allow users to edit the manifest directory location and also to be able to rename the manifest.

## List all the issues related to this PR

- [ ] New Feature 
- [ ] Enhancement 
- [ ] Bug fix
- [ ] Tests
- [ ] Documentation
- [ ] CI/CD 

## What environments does this code impact?

- [ ] Automation (CI, tools, etc)
- [ ] Cloud
- [ ] Operator Managed Deployments
- [x] None

## How was this code tested?

- [ ] assisted-test-infra environment
- [ ] dev-scripts environment
- [ ] Reviewer's test appreciated
- [ ] Waiting for CI to do a full test run
- [ ] Manual (Elaborate on how it was tested)
- [x] No tests needed

## Checklist

- [ ] Title and description added to both, commit and PR.
- [ ] Relevant issues have been associated (see [CONTRIBUTING] guide)
- [ ] This change does not require a documentation update (docstring, `docs`, README, etc)
- [ ] Does this change include unit-tests (note that code changes require unit-tests)

## Reviewers Checklist

- Are the title and description (in both PR and commit) meaningful and clear?
- Is there a bug required (and linked) for this change?
- Should this PR be backported?

[Kubernetes community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md

### CODE REVIEW

File: swagger.yaml [hunk 1]
```diff
@@ -2930,6 +2930,60 @@ paths:
...
+          format: uuid
+          required: true
+        - in: body
+          name: UpdateManifestParams
+          description: The manifest to be updated.
+          required: true
+          schema:
+            $ref: '#/definitions/update-manifest-params'
+      responses:
+        "201":
```
Comment (avishayt) [inline swagger.yaml:2954]: Should be 200 (OK) not 201 (Created)

File: swagger.yaml [hunk 2]
```diff
@@ -6256,6 +6310,39 @@ definitions:
       - file_name
       - content
 
+  update-manifest-params:
+    type: object
+    properties:
+      src_folder:
+        description: The source folder. Manifests can be placed in 'manifests' or 'openshift' directories.
```
Comment (avishayt) [inline swagger.yaml:6317]:
```suggestion
      folder:
        description: The folder for the manifest to modify.
```

File: swagger.yaml [hunk 3]
```diff
@@ -6256,6 +6310,39 @@ definitions:
...
+        type: string
+        enum: [manifests,openshift]
+        default: manifests
+        x-nullable: false
+      src_file_name:
+        description: The name of the manifest to customize the installed OCP cluster.
```
Comment (avishayt) [inline swagger.yaml:6323]:
```suggestion
      file_name:
        description: The file name for the manifest to modify.
```

File: swagger.yaml [hunk 4]
```diff
@@ -6256,6 +6310,39 @@ definitions:
...
+        type: string
+        pattern: '^[^/]*\.(yaml|yml|json)$'
+        x-nullable: false
+      dest_folder:
+        description: The new folder that will contain the files. Manifests can be placed in 'manifests' or 'openshift' directories.
```
Comment (avishayt) [inline swagger.yaml:6328]:
```suggestion
      updated_folder:
        description: The new folder for the manifest. Manifests can be placed in 'manifests' or 'openshift' directories.
```

File: swagger.yaml [hunk 5]
```diff
@@ -6256,6 +6310,39 @@ definitions:
...
+        type: string
+        enum: [manifests,openshift]
+        default: manifests
+        x-nullable: true
+      dest_file_name:
+        description: The name that the file is to be renamed to
```
Comment (avishayt) [inline swagger.yaml:6334]:
```suggestion
      updated_file_name:
        description: The new file name for the manifest.
```

File: swagger.yaml [hunk 6]
```diff
@@ -6256,6 +6310,39 @@ definitions:
...
+        type: string
+        pattern: '^[^/]*\.(yaml|yml|json)$'
+        x-nullable: true
+      content:
+        description: base64 encoded manifest content.
```
Comment (avishayt) [inline swagger.yaml:6339]:
```suggestion
      updated_content:
        description: The new base64 encoded manifest content.
```

File: internal/manifests/manifests.go [hunk 7]
```diff
@@ -147,50 +131,86 @@ func (m *Manifests) DeleteClusterManifestInternal(ctx context.Context, params op
...
-		params.Folder = &defaultFolder
+	log.Infof("Done deleting cluster manifest %s for cluster %s", path, params.ClusterID.String())
+	return nil
+}
+
+func (m *Manifests) UpdateClusterManifestInternal(ctx context.Context, params operations.V2UpdateClusterManifestParams) (*models.Manifest, error) {
+	if params.UpdateManifestParams.UpdatedFolder == nil {
+		params.UpdateManifestParams.UpdatedFolder = &params.UpdateManifestParams.Folder
+		params.UpdateManifestParams.UpdatedFileName = &params.UpdateManifestParams.FileName
 	}
```
Comment (avishayt) [inline internal/manifests/manifests.go:162]: What if I only pass UpdatedFolder or only pass UpdatedFileName?  Both of these cases will break

File: internal/manifests/manifests.go [hunk 8]
```diff
@@ -56,58 +57,41 @@ func (m *Manifests) CreateClusterManifestInternal(ctx context.Context, params op
...
+	folder, fileName, path := m.getManifestPathsFromParameters(ctx, params.CreateManifestParams.Folder, params.CreateManifestParams.FileName)
+	log.Infof("Folder = '%s' and filename = '%s' and path = '%s'", folder, fileName, path)
 
 	// Verify that the manifests are created for a valid cluster
 	// to align kube-api and ocm behavior.
 	// In OCM, this is validated at the authorization layer. In other
 	// authorization scheme, it does not and therefore should be checked
 	// at the application level.
 	if !cluster.ClusterExists(m.db, params.ClusterID) {
-		return nil, common.NewApiError(http.StatusNotFound, fmt.Errorf("Object Not Found"))
```
Comment (avishayt) [inline internal/manifests/manifests.go:70]: I believe the original "Object Not Found" message was purposefully created to align with the same error message returned in other cases.  It should be left as is.

File: internal/manifests/manifests.go [hunk 9]
```diff
@@ -121,7 +105,7 @@ func (m *Manifests) ListClusterManifestsInternal(ctx context.Context, params ope
...
```
Comment (avishayt) [inline internal/manifests/manifests.go:124]: Same as above

File: internal/manifests/manifests.go [hunk 10]
```diff
@@ -147,50 +131,86 @@ func (m *Manifests) DeleteClusterManifestInternal(ctx context.Context, params op
...
 
-	// This call both verifies that the manifests are created for a valid cluster
-	// to align kube-api and ocm behavior, and get the cluster object to check that
-	// it is in valid state.
 	cluster, err := common.GetClusterFromDB(m.db, params.ClusterID, common.SkipEagerLoading)
 	if err != nil {
-		return common.NewApiError(http.StatusNotFound, fmt.Errorf("Object Not Found"))
+		err = fmt.Errorf("could not find cluster with id %s", params.ClusterID.String())
+		log.Infof(err.Error())
+		return common.NewApiError(http.StatusNotFound, err)
```
Comment (avishayt) [inline internal/manifests/manifests.go:138]: Same as above - please revert

File: internal/manifests/manifests.go [hunk 11]
```diff
@@ -147,50 +131,86 @@ func (m *Manifests) DeleteClusterManifestInternal(ctx context.Context, params op
...
-	objectName := GetManifestObjectName(params.ClusterID, fileName)
-	exists, err := m.objectHandler.DoesObjectExist(ctx, objectName)
+	srcFolder, srcFileName, srcPath := m.getManifestPathsFromParameters(ctx, &params.UpdateManifestParams.Folder, &params.UpdateManifestParams.FileName)
+	destFolder, destFileName, destPath := m.getManifestPathsFromParameters(ctx, params.UpdateManifestParams.UpdatedFolder, params.UpdateManifestParams.UpdatedFileName)
+
+	cluster, err := common.GetClusterFromDB(m.db, params.ClusterID, common.SkipEagerLoading)
 	if err != nil {
-		log.WithError(err).Errorf("Failed to delete cluster manifest %s", objectName)
-		return common.NewApiError(http.StatusInternalServerError, err)
+		return nil, m.prepareAndLogError(ctx, http.StatusNotFound, fmt.Errorf("could not find cluster with id %s", params.ClusterID.String()))
```
Comment (avishayt) [inline internal/manifests/manifests.go:168]: Return the same error as above

File: internal/manifests/manifests.go (hunk 7 above)
Comment (avishayt) [reply internal/manifests/manifests.go:162]: Please add unit tests for these cases

**Review by avishayt — APPROVED** [review]

**Review by osherdp — APPROVED** [review]

**Pull request merged**